server/src/main/java/org/opensearch/search/resource_limit_group/tracker/ResourceLimitsGroupResourceUsageTrackerService.java

```diff
@@ -0,0 +1,300 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+
+/**
+ * This class tracks requests per resourceLimitGroups
+ */
+public class ResourceLimitsGroupResourceUsageTrackerService extends ResourceLimitGroupTaskCanceller
+    implements
+        TaskManager.TaskEventListeners,
+        ResourceLimitGroupResourceUsageTracker,
+        ResourceLimitGroupPruner {
+
+    private static final String CPU = "cpu";
+    private static final String JVM = "jvm";
+    private static final List<String> TRACKED_RESOURCES = List.of(JVM);
+    private static final int numberOfAvailableProcessors = Runtime.getRuntime().availableProcessors();
+    private static final long totalAvailableJvmMemory = Runtime.getRuntime().totalMemory();
+    private final LongSupplier timeNanosSupplier;
+    /**
+     * ResourceLimitGroup ids which are marked for deletion in between the
+     * {@link org.opensearch.search.resource_limit_group.ResourceLimitGroupService} runs
+     */
+    private List<String> toDeleteResourceLimitGroups;
+    private List<ResourceLimitGroup> activeResourceLimitGroups;
+    /**
+     * This var will hold the sandbox level resource usage as per last run of
+     * {@link org.opensearch.search.resource_limit_group.ResourceLimitGroupService}
+     */
+    private Map<String, Map<String, Double>> resourceUsage;
+    /**
+     * We are keeping this as instance member as we will need this to identify the contended resource limit groups
+     * resourceLimitGroupName -> List&lt;ResourceLimitGroupTask&gt;
+     */
+    private Map<String, List<Task>> resourceLimitGroupTasks;
+    private final TaskManager taskManager;
+    private final TaskResourceTrackingService taskResourceTrackingService;
+    private final ClusterService clusterService;
+    private final CancellableTaskSelector taskSelector;
+
+    /**
+     * SandboxResourceTrackerService constructor
+     * @param taskManager
+     * @param taskResourceTrackingService
+     * @param clusterService
+     */
+    @Inject
+    public ResourceLimitsGroupResourceUsageTrackerService(
+        final TaskManager taskManager,
+        final TaskResourceTrackingService taskResourceTrackingService,
+        final ClusterService clusterService,
+        final LongSupplier timeNanosSupplier,
+        final CancellableTaskSelector taskSelector
+    ) {
+        super(taskSelector);
+        this.taskManager = taskManager;
+        this.taskResourceTrackingService = taskResourceTrackingService;
+        toDeleteResourceLimitGroups = new ArrayList<>();
+        this.clusterService = clusterService;
+        this.timeNanosSupplier = timeNanosSupplier;
+        this.taskSelector = taskSelector;
+    }
+
+    @Override
+    public void updateResourceLimitGroupsResourceUsage() {
+        activeResourceLimitGroups = new ArrayList<>(clusterService.state().metadata().resourceLimitGroups().values());
```

Comment:
should we avoid storing these values at the instance level to avoid synchronization issues and always fetch the activeResourceLimitGroups from the clusterService ?

Reply:
This is calculated every second so this is always upto date. Am I missing something here ?